solr/core/src/java/org/apache/solr/servlet/HttpSolrProxy.java

```diff
@@ -0,0 +1,148 @@
+
+  // Methods that shouldn't have a body according to HTTP spec
+  private static final Set<String> NO_BODY_METHODS = Set.of("GET", "HEAD", "DELETE");
+
+  static void doHttpProxy(
+      HttpClient httpClient,
+      HttpServletRequest servletReq,
+      HttpServletResponse servletRsp,
+      String url)
+      throws Throwable {
+    Request proxyReq = httpClient.newRequest(url).method(servletReq.getMethod());
+
+    // clearing them first to ensure there's no stock entries (e.g. user-agent)
+    proxyReq.headers(proxyFields -> copyRequestHeaders(servletReq, proxyFields.clear()));
+
+    if (!NO_BODY_METHODS.contains(servletReq.getMethod())) {
+      proxyReq.body(
+          new InputStreamRequestContent(servletReq.getContentType(), servletReq.getInputStream()));
+    }
+
+    CompletableFuture<Result> resultFuture = new CompletableFuture<>();
+
+    proxyReq.send(
+        new Response.Listener() {
+          private final byte[] buffer = new byte[8192];
+
+          @Override
+          public void onBegin(Response response) {
+            servletRsp.setStatus(response.getStatus());
+          }
+
+          @Override
+          public void onHeaders(Response response) {
+            copyResponseHeaders(response, servletRsp);
+          }
+
+          @Override
+          public void onContent(Response response, ByteBuffer content) {
+            try {
+              final OutputStream clientOutputStream = servletRsp.getOutputStream();
+
+              // Copy content to the client's output stream in chunks using the existing buffer
+              int remaining = content.remaining();
+              while (remaining > 0) {
+                int chunkSize = Math.min(remaining, buffer.length);
+                content.get(buffer, 0, chunkSize);
+                clientOutputStream.write(buffer, 0, chunkSize);
+                remaining -= chunkSize;
+              }
+              clientOutputStream.flush();
+            } catch (IOException e) {
+              throw new RuntimeException(e);
+            }
+          }
+
+          @Override
+          public void onComplete(Result result) {
+            resultFuture.complete(result);
+          }
+        });
+
+    Result result = resultFuture.get(); // waits
+    var failure = result.getFailure();
+    if (failure != null) {
+      throw failure;
+    }
+  }
+
+  private static void copyRequestHeaders(
+      HttpServletRequest servletReq, HttpFields.Mutable proxyFields) {
+    servletReq
+        .getHeaderNames()
+        .asIterator()
+        .forEachRemaining(
+            headerName -> {
+              HttpHeader knownHeader = HttpHeader.CACHE.get(headerName); // maybe null
+              if (!HOP_BY_HOP_HEADERS.contains(knownHeader)) {
+                // NOCOMMIT see SOLR-7274 SPNEGO support:
+                // && HttpHeader.HOST != knownHeader
+                // && HttpHeader.AUTHORIZATION != knownHeader
+                // && HttpHeader.ACCEPT != knownHeader) {
```

Comment:
@anshumg, 10 years ago in [SOLR-17286](https://issues.apache.org/jira/browse/SOLR-17286),  you committed this particular exception list of 3 headers to not copy on the grounds of SPNEGO.  In the comments, it was actually @chatman who pointed this out.  I don't understand it; my thinking is that a proxy should be a proxy and pass everything through except for the hop-by-hop header list according to RFC.  If I restore these exceptions, then the wonderful test recently added will fail: https://github.com/apache/solr/blob/eaceca8911ef06c89a293f2437a6a560f34b1a43/solr/modules/jwt-auth/src/test/org/apache/solr/security/jwt/JWTAuthPluginIntegrationTest.java#L301
(see PR #3397 that added it)